include/bit_algorithm_details.hpp

```diff
@@ -1,29 +1,96 @@
-// ======================== BIT_ALGORITHM_DETAILS =========================== //
-// Project:     The Experimental Bit Algorithms Library
-// Name:        iterator_utils.hpp
-// Description: Implementation of algorithm-related utilities not present in 
-// 		        algorithms header.
-// Creator:     Vincent Reverdy
-// Contributor: Bryce Kille [2019]
-//              Collin Gress [2019]
-// License:     BSD 3-Clause License
+// ========================= BIT ALGORITHM DETAILS  ========================= //
+// Project: The Experimental Bit Algorithms Library
+// Name: bit_algorithm_details.hpp
+// Description: A set of utilities to assist in writing algorithms 
+// Creator: Vincent Reverdy
+// Contributor(s): Vincent Reverdy [2019]
+//                 Collin Gress [2019]
+//                 Bryce Kille [2019]
+// License: BSD 3-Clause License
 // ========================================================================== //
 #ifndef _BIT_ALGORITHM_DETAILS_HPP_INCLUDED
 #define _BIT_ALGORITHM_DETAILS_HPP_INCLUDED
 // ========================================================================== //
 
 
-
-// ================================ PREAMBLE ================================ //
+// ============================== PREAMBLE ================================== //
 // C++ standard library
 // Project sources
-#include "bit.hpp"
 // Third-party libraries
 // Miscellaneous
-namespace bit {
-// ========================================================================== //
 
+namespace bit {
+/*
+ * Takes two words and merges them together.
+ * 
+ * @param w1 the first operand of the merge
+ * @param w2 the second operand of the merge
+ * @param w1_position the position within w1 to start the merge from
+ * @return a word given by merging w1 and w2
+ */ 
+template <class T>
+T _shift_and_merge(T w1, T w2, std::size_t w1_position) {
+
+  if (w1_position == 0) {
+    return w1;
+  }
+
+  constexpr std::size_t num_digits = bit::binary_digits<T>::value;
+
+  T tmp1 = w1 << w1_position;
+  T tmp2 = w2 >> (num_digits - w1_position);
+  return tmp1 | tmp2;
+}
 
+/*
+ * This function reads a T's worth of bits starting from wherever
+ * the start bit_iterator is pointing. This means that we might be making
+ * a word read from words which span multiple words.
+ *
+ * @param start the bit_iterator to start reading from
+ * @param guard the furthest position to read to, ensuring there are not
+ * memory access violations
+ * @return a word read starting from the start position
+ */
+template <class WrappedIter, class T = typename WrappedIter::word_type>
+T _read_word_raw(WrappedIter start, WrappedIter guard) {
+  const unsigned long dist = std::distance(start, guard);
```

Comment:
Maybe I am misremembering, but I thought at the last meeting we discussed that `distance` might not be able to be used on `InputIt`. Not only this, but this function still only handles the case where we are spanning at most two words. 

I think it's worth discussing at our next meeting. While we can have general logic for reading any `T1` bits from any `T2` bit iterator, it may prove faster if we know `T1` and `T2` are the same, then we know that we will need to read from at most 2 words.  If that is the case, though, we should decide on some naming and perhaps overload `_read_word_raw`